src/main/scala/cromwell/webservice/CallCachingParameters.scala

```diff
@@ -0,0 +1,126 @@
+package cromwell.webservice
+
+import cromwell.engine.db.DataAccess._
+import cromwell.engine.db.{DataAccess, ExecutionDatabaseKey}
+import cromwell.engine.{ExecutionStatus, WorkflowId}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+import scalaz.Scalaz._
+import scalaz.{Failure, Success, ValidationNel}
+
+case class CallCachingParameters private(workflowId: WorkflowId, callKey: Option[ExecutionDatabaseKey], allow: Boolean)
+
+object CallCachingParameters {
+
+  private [webservice] def validateRecognizedKeys(rawParameters: Seq[(String, String)]): ValidationNel[String, Unit] = {
+    val badKeys = rawParameters collect { case (k, _) if k.toLowerCase != "allow" => k }
+    if (badKeys.nonEmpty) ("Found unrecognized keys: " + badKeys.mkString(", ")).failureNel else ().successNel
+  }
+
+  private [webservice] def validateAllow(rawParameters: Seq[(String, String)]): ValidationNel[String, Boolean] = {
+    val allows = rawParameters.collect({ case x if x._1.toLowerCase == "allow" => x._2 }).toSet
+    val (booleans, nonBooleans) = allows map { a => (a, Try(a.toBoolean)) } partition { _._2.isSuccess }
+
+    val allBooleans = if (nonBooleans.nonEmpty) {
+      val values = (nonBooleans map { _._1 } ).mkString(", ")
+      s"Found non-boolean 'allow' values: $values".failureNel
+    } else {
+      ().successNel
+    }
+
+    val (trues, falses) = booleans partition { _._2.get }
+    val coherentValues = if (trues.nonEmpty && falses.nonEmpty) "Found both true and false 'allow' values".failureNel else trues.nonEmpty.successNel
+    (allBooleans |@| coherentValues) {
+      case (_, allow) => allow
+    }
+  }
+
+
+  private def validateWorkflowExists(workflowId: WorkflowId, dataAccess: DataAccess)(implicit ec: ExecutionContext): Future[ValidationNel[String, Unit]] = {
+    dataAccess.getWorkflowState(workflowId) map {
+      case Some(_) => ().successNel
+      case None => s"Workflow not found: ${workflowId.id.toString}".failureNel
+    }
+  }
+
+  private [webservice] def validateCallName(callName: Option[String]): ValidationNel[String, Option[ExecutionDatabaseKey]] = {
+    callName map { name =>
+      val CallNamePattern = "((?:[a-zA-Z][a-zA-Z0-9_]*)\\.(?:[a-zA-Z][a-zA-Z0-9_]*))(?:\\.(\\d+))?".r
```

Comment:
This might be worth extracting to put at the object level so we only declare it once.
Also, even if the regexp is relatively easy to read, I think a simple comment with an example above the declaration is helpful for regexps.